### DISCUSSION TITLE
Single-fire reliable events

### DISCUSSION THREAD

**Discussion by bryanmylee** [foxssake/netfox#189] [Q&A] [answered]
Right now, `RollbackSynchronizer` keeps state replicated across the network.

However, for single-process-frame events like `Input.is_action_just_pressed("jump")`, Netfox does not react to the state change / event.

I need the event to be triggered **only once** right after the frame that it was called in, but I can't find any solutions at the moment.

**Comment by elementbound** [comment, marked as answer]
This is caused by the custom tick loop. From the Input docs:

> Returns true when the user has started pressing the action event in the current frame or physics tick.

What happens is between pressing the button, a frame might pass before the tick loop runs next. Imagine a timeline like this:

1. _process
2. Input is pressed
3. _process **<- is_action_just_pressed is true**
4. _process **<- is_action_just_pressed is false**
5. NetworkTime.on_tick **<- is_action_just_pressed is false**
4. _process **<- is_action_just_pressed is false**

---

You can move your input to a separate variable on your class, gather input on every frame, use the gathered input in the first tick in the loop, then reset the input. Something like this:

```gdscript
var is_jumping: bool = false

func _ready():
    NetworkTime.after_tick.connect(_reset)

func _process():
    if Input.is_action_just_pressed("jump"):
        is_jumping = true

func _reset(_delta, _tick):
    is_jumping = false
```

This way, if the input is pressed between two ticks, it will be remembered. After the first tick in the loop, the input is reset back to false. I'm mentioning the tick loop, because in some cases the tick loop needs to run multiple ticks to catch up. So in this case, even if the input was pressed, it will be used only in the first tick of the loop. 

Resetting only after the tick loop can also be useful for continuous inputs like movement.

Let me know if this clears it up!

**Reply by bryanmylee** [reply]
Ah that's perfect! I forgot `NetworkTime.after_tick` was a thing, that lets me simulate `is_action_just_pressed` on the network tick instead of the process tick!

Thanks for all your great work!

**Comment by bryanmylee** [comment]
For anyone who's trying to do this same thing in the future, I'm not exactly sure why, but make sure to clear input in `after_tick_loop`, not `after_tick`.

**Reply by elementbound** [reply]
Interesting, I'll look into that. Thanks for the heads-up, appreciated!
